src/hssm/hssm.py

```diff
@@ -592,35 +614,120 @@ def sample_posterior_predictive(
                     + "Please either provide an idata object or sample the model first."
                 )
             idata = self._inference_obj
+            _logger.info(
+                "idata=None, we use the traces assigned to the HSSM object as idata."
+            )
+
+        if "posterior_predictive" in idata.groups():
+            if idata is not None:
+                del idata["posterior_predictive"]
+                _logger.warning(
+                    "pre-existing posterior_predictive group deleted from idata. \n"
+                )
 
         if self._check_extra_fields(data):
             self._update_extra_fields(data)
 
-        if n_samples is not None:
-            # Make a copy of idata, set the `posterior` group to be a random sub-sample
-            # of the original (draw dimension gets sub-sampled)
-            idata_copy = idata.copy()
-            idata_random_sample = _random_sample(
-                idata_copy["posterior"], n_samples=n_samples
+        if isinstance(draws, np.ndarray):
+            draws = draws.astype(int)
+        elif isinstance(draws, list):
+            draws = np.array(draws).astype(int)
+        elif isinstance(draws, int | float):
+            draws = np.arange(int(draws))
+        elif draws is None:
+            draws = idata["posterior"].draw.values
+        else:
+            raise ValueError(
+                "draws must be an integer, " + "a list of integers, or a numpy array."
             )
-            delattr(idata_copy, "posterior")
-            idata_copy.add_groups(posterior=idata_random_sample)
 
-            # If the user specifies an inplace operation, we need to modify the original
-            if inplace:
+        assert isinstance(draws, np.ndarray)
+
+        # Make a copy of idata, set the `posterior` group to be a random sub-sample
+        # of the original (draw dimension gets sub-sampled)
+
+        idata_copy = idata.copy()
+
+        if (draws.shape != idata["posterior"].draw.values.shape) or (
+            (draws.shape == idata["posterior"].draw.values.shape)
+            and not np.allclose(draws, idata["posterior"].draw.values)
+        ):
+            # Reassign posterior to sub-sampled version
+            setattr(idata_copy, "posterior", idata["posterior"].isel(draw=draws))
```

Comment:
Are there any differences between `setattr()` and `idata.add_groups()`?

Reply:
to be honest I don't know... let me look into that independently to understand it properly.

Reply:
Actually, at least used somewhat semantically here, `add_groups` is about new groups, `setattr` is about reassigning to pre-existing group.